pkg/oci/filestorage/filestorage.go

```diff
@@ -0,0 +1,266 @@
+	if err != nil {
+		return flexvolume.Fail(err.Error())
+	}
+
+	if len(privateIps) == 0 {
+		return flexvolume.Fail("MountRarget:% has zero private IPs", *mountTarget.Id)
+	}
+
+	var path = fmt.Sprintf("/mnt/%s/%s", *privateIps[0].IpAddress, filesystemOCID)
+
+	err = c.DetachFileSystemToMountTarget(filesystem, mountTarget, path)
+	if err != nil {
+		return flexvolume.Fail(err.Error())
+	}
+
+	return flexvolume.Succeed("Detach %s from %s", *filesystem.Id, *mountTarget.Id)
+}
+
+// WaitForAttach does nothing but return true as we have done the
+// wait in the mountdevice calls and this means that no creds are needed on worker nodes
+func (d OCIFilestorageDriver) WaitForAttach(mountDevice string, _ flexvolume.Options) flexvolume.DriverStatus {
+	return flexvolume.DriverStatus{
+		Status: flexvolume.StatusSuccess,
+		Device: mountDevice,
+	}
+}
+
+// IsAttached checks whether the volume is attached to the host.
+func (d OCIFilestorageDriver) IsAttached(opts flexvolume.Options, nodeName string) flexvolume.DriverStatus {
+	c, err := client.New(common.GetConfigPath())
+	if err != nil {
+		return flexvolume.Fail(err.Error())
+	}
+
+	filesystemOCID := opts["kubernetes.io/pvOrVolumeName"]
+
+	filesystem, err := c.GetFileSystem(filesystemOCID)
+	if err != nil {
+		log.Printf("Failed to get FileSystem")
+		return flexvolume.Fail(err.Error())
+	}
+
+	mountTarget, err := c.GetMountTargetForAD(*filesystem.AvailabilityDomain)
+	if err != nil {
+		log.Printf("Failed to GetMountTargetForAD")
+		return flexvolume.Fail(err.Error())
+	}
+
+	privateIps, err := c.GetMountTargetIPS(mountTarget)
+	if err != nil {
+		log.Printf("Failed GetMountTargetIPS")
+		return flexvolume.Fail(err.Error())
+	}
+	log.Printf("Mount TargetIPS:%v", privateIps)
+
+	if len(privateIps) == 0 {
+		return flexvolume.Fail("MountRarget:% has zero private IPs", *mountTarget.Id)
+	}
+
+	var path = fmt.Sprintf("/mnt/%s/%s", *privateIps[0].IpAddress, filesystemOCID)
+
+	attached, err := c.IsFileSystemAttached(filesystem, mountTarget, path)
+	if err != nil {
+		log.Printf("Failed IsFileSystemAttached")
+		return flexvolume.Fail(err.Error())
+	}
+	return flexvolume.DriverStatus{
+		Status:   flexvolume.StatusSuccess,
+		Attached: attached,
+	}
+}
+
+// MountDevice connects the iSCSI target on the k8s worker node before mounting
+// and (if necessary) formatting the disk.
+func (d OCIFilestorageDriver) MountDevice(mountDir, mountDevice string, opts flexvolume.Options) flexvolume.DriverStatus {
+
+	parts := strings.Split(mountDevice, "/")
+	IPaddress := parts[2]
+	source := fmt.Sprintf("%s:%s", IPaddress, mountDevice)
+
+	mounter := mount.New("")
+
+	notMnt, err := mounter.IsLikelyNotMountPoint(mountDir)
+	log.Printf("NFS mount set up: %s %v %v", mountDir, !notMnt, err)
+	if err != nil && !os.IsNotExist(err) {
+		return flexvolume.Fail(err.Error())
+	}
+	if !notMnt {
+		return flexvolume.Fail("Already mount point")
+	}
+	os.MkdirAll(mountDir, 0750)
+
+	options := []string{}
+	/*if b.readOnly {
```

Comment:
Remove?

Reply:
Also missing support for read only here. 

```
if opts[flexvolume.OptionReadWrite] == "ro" {
		options = append(options, "ro")
}
```